src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -865,7 +872,14 @@ private void VisitSwitchExpressionCore(BoundSwitchExpression node, bool inferTyp
                 SetState(getStateForArm(arm, labelStateMap));
                 // https://github.com/dotnet/roslyn/issues/35836 Is this where we want to take the snapshot?
                 TakeIncrementalSnapshot(arm);
-                VisitPatternForRewriting(arm.Pattern);
+
+                if (!State.Reachable && arm.WhenClause != null)
+                {
+                    // Unreachable when clauses are not visited in `LearnFromDecisionDag`.
+                    VisitForRewriting(arm.WhenClause);
+                }
+
+                VisitForRewriting(arm.Pattern);
```

Comment:
It may not make an observable difference, but it makes the most sense to me to visit the WhenClause after the Pattern. I think of the when-clause as being evaluated only after the pattern is found to match.

Reply:
That makes sense, thanks. Moved.